File: python/day_18b.py
#!/usr/bin/python3
from collections import deque

def getExpr(line):
    expr = deque()
    for char in line:
        if char != ' ':
            expr.append(char)
    return expr

def getSubExpr(expr):
    bracket_count = 0
    expr.popleft()
    new_expr = deque()
    while bracket_count > 0 or expr[0] != ')':
        if expr[0] == '(':
            bracket_count += 1
        elif expr[0] == ')':
            bracket_count -= 1
        new_expr.append(expr[0])
        expr.popleft()
    expr.popleft()
    return new_expr

def calcExprValue(expr):
    while len(expr) > 1:
        operand1 = int()
        operand2 = int()
        if expr[0] != '(':
            operand1 = int(expr.popleft())
        else:
            operand1 = calcExprValue(getSubExpr(expr))
        operator = expr.popleft()
        if expr[0] != '(':
            operand2 = int(expr.popleft())
        else:
            operand2 = calcExprValue(getSubExpr(expr))
        if operator == '+':
            expr.appendleft(str(operand1 + operand2))
        elif operator == '*':
            expr.appendleft(str(operand1 * operand2))
    return int(expr[0])

def calcExprValueAdv(expr):
    simplifiedEquation = deque()
    while len(expr) > 1:
        operand1 = int()
        operand2 = int()
        if expr[0] != '(':
            operand1 = int(expr.popleft())
        else:
            operand1 = calcExprValueAdv(getSubExpr(expr))
        operator = expr.popleft()
        if expr[0] != '(':
            operand2 = int(expr.popleft())
        else:
            operand2 = calcExprValueAdv(getSubExpr(expr))

        if operator == '+':
            expr.appendleft(str(operand1 + operand2))
        elif operator == '*':
            expr.appendleft(str(operand2))
            simplifiedEquation.append(str(operand1))
            simplifiedEquation.append(operator)
    simplifiedEquation.append(expr.popleft())
    return calcExprValue(simplifiedEquation)


def main():
    f = open("../cpp/day_18_data.txt")
    sum = 0
    for line in f:
        line = line.strip('\n')
        expr = getExpr(line)
        sum += calcExprValueAdv(expr)
    print(sum)
    return sum

if __name__ == "__main__":
    main()

File: cpp/day_18b.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <queue>
#include <string>
#include <vector>

std::deque<std::string> convertStringToDeque(const std::string& line) {
  std::deque<std::string> cur_expr;
  for (int i = 0; i < line.size(); i++) {
    if (line[i] != ' ') {
      cur_expr.push_back(std::string{line[i]});
    }
  }
  return cur_expr;
}

std::deque<std::string> getSubExpr(std::deque<std::string>& expr) {
  int brack_count = 0;
  expr.pop_front();
  std::deque<std::string> new_expr;
  while (brack_count > 0 || expr.front() != ")") {
    if (expr.front() == "(") {
      brack_count++;
    }
    if (expr.front() == ")") {
      brack_count--;
    }
    new_expr.push_back(expr.front());
    expr.pop_front();
  }
  expr.pop_front();
  return new_expr;
}

long long calcExpressionValue(std::deque<std::string>& expr) {
  while (expr.size() > 1) {
    long long operand1;
    long long operand2;
    std::string op;

    if (expr.front() != "(") {
      operand1 = std::stol(expr.front());
      expr.pop_front();
    } else {
      auto new_expr = getSubExpr(expr);
      operand1 = calcExpressionValue(new_expr);
    }

    op = expr.front();
    expr.pop_front();

    if (expr.front() != "(") {
      operand2 = std::stol(expr.front());
      expr.pop_front();
    } else {
      auto new_expr = getSubExpr(expr);
      operand2 = calcExpressionValue(new_expr);
    }
    if (op == "+") expr.push_front(std::to_string(operand1 + operand2));
    if (op == "*") expr.push_front(std::to_string(operand1 * operand2));
  }
  return std::stol(expr.front());
}

long long calcExpressionValueAdvanced(std::deque<std::string>& expr) {
  std::deque<std::string> simplifiedEquation;
  while (expr.size() > 1) {
    long long operand1;
    long long operand2;
    std::string op;

    if (expr.front() != "(") {
      operand1 = std::stol(expr.front());
      expr.pop_front();
    } else {
      auto new_expr = getSubExpr(expr);
      operand1 = calcExpressionValueAdvanced(new_expr);
    }

    op = expr.front();
    expr.pop_front();

    if (expr.front() != "(") {
      operand2 = std::stol(expr.front());
      expr.pop_front();
    } else {
      auto new_expr = getSubExpr(expr);
      operand2 = calcExpressionValueAdvanced(new_expr);
    }
    if (op == "+") expr.push_front(std::to_string(operand1 + operand2));
    // If the operation is multiplcation, add the second operand back to the
    // front of the original deque and add the first operand the the operator to
    // the simplified equation to build up an equation that consists of only
    // multiplication
    if (op == "*") {
      expr.push_front(std::to_string(operand2));
      simplifiedEquation.push_back(std::to_string(operand1));
      simplifiedEquation.push_back(op);
    }
  }
  // Get the final operand from the equation into simplified equation
  std::string finalOperand = expr.front();
  simplifiedEquation.push_back(finalOperand);
  expr.pop_front();
  std::cout << "Hello" << '\n';
  // Finally solve the remaining multiplication using our original
  // solveEquation()
  return calcExpressionValue(simplifiedEquation);
}

int main() {
  std::fstream file{"../input/day_18_input"};
  std::string line;
  long long sum = 0;
  while (std::getline(file, line)) {
    line.erase(std::remove_if(std::begin(line), std::end(line),
                              [](auto c) { return !isprint(c) || c == ' '; }),
               std::end(line));
    std::cout << line << '\n';
    auto d_line = convertStringToDeque(line);
    sum += calcExpressionValueAdvanced(d_line);
  }

  std::cout << sum << '\n';
  return sum;
}
